lib/internal/modules/esm/get_format.js

```diff
@@ -85,42 +83,66 @@ function underNodeModules(url) {
 
 /**
  * @param {URL} url
- * @param {{parentURL: string}} context
+ * @param {{parentURL: string; source?: Buffer}} context
  * @param {boolean} ignoreErrors
  * @returns {string}
  */
-function getFileProtocolModuleFormat(url, context, ignoreErrors) {
+function getFileProtocolModuleFormat(url, context = { __proto__: null }, ignoreErrors) {
+  const { source } = context;
   const ext = extname(url);
 
   if (ext === '.js') {
     const packageType = getPackageType(url);
     if (packageType !== 'none') {
       return packageType;
     }
+
     // The controlling `package.json` file has no `type` field.
-    if (defaultType === 'module') {
-      // An exception to the type flag making ESM the default everywhere is that package scopes under `node_modules`
-      // should retain the assumption that a lack of a `type` field means CommonJS.
-      return underNodeModules(url) ? 'commonjs' : 'module';
+    switch (getOptionValue('--experimental-default-type')) {
+      case 'module': { // The user explicitly passed `--experimental-default-type=module`.
+        // An exception to the type flag making ESM the default everywhere is that package scopes under `node_modules`
+        // should retain the assumption that a lack of a `type` field means CommonJS.
+        return underNodeModules(url) ? 'commonjs' : 'module';
+      }
+      case 'commonjs': { // The user explicitly passed `--experimental-default-type=commonjs`.
+        return 'commonjs';
+      }
+      default: { // The user did not pass `--experimental-default-type`.
+        // `source` is undefined when this is called from `defaultResolve`;
+        // but this gets called again from `defaultLoad`/`defaultLoadSync`.
+        if (source && getOptionValue('--experimental-detect-module')) {
+          return containsModuleSyntax(`${source}`, fileURLToPath(url)) ? 'module' : 'commonjs';
```

Comment:
This only supports string source, that's a bit annoying. Why can't we pass a `Uint8Array` to the C++ binding?

Reply:
I'm not opposed if you can get that to work. `compileFunction` expects a string though.